pkg/oci/routes.go

```diff
@@ -0,0 +1,291 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"context"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/cloudprovider"
+
+	"strings"
+
+	"github.com/golang/glog"
+	"github.com/oracle/oci-cloud-controller-manager/pkg/oci/util"
+	"github.com/oracle/oci-go-sdk/core"
+	"github.com/pkg/errors"
+)
+
+var _ cloudprovider.Routes = &CloudProvider{}
+
+/*
+OCI doesn't have distinct add/remove/update route commands, only the ability to
+update the whole RouteTable.
+
+K8s dispatches up to 200 concurrent creates.
+
+As a result ADD/Create races with itself, especially in light of DELETE.
+
+Ultimately the reconciliation restores order in the galaxy, but only after some
+disruptions in the force.
+
+One _could_ add some locking on the Get/Update of RouteTable, but it's not
+clear that should be required.
+*/
+
+// CreateRoute for OCI, we have to udpate the whole RouteTable not just add the singular route
+func (cp *CloudProvider) CreateRoute(ctx context.Context, clusterName string, nameHint string, route *cloudprovider.Route) error {
+	glog.V(6).Info("Adding route to route table: ", route)
+
+	node, err := cp.NodeLister.Get(string(route.TargetNode))
+
+	ocid := util.MapProviderIDToInstanceID(node.Spec.ProviderID)
+	vnic, err := cp.client.Compute().GetPrimaryVNICForInstance(ctx, "", ocid)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	ipOcid, err := cp.client.Networking().GetOCIDFromIP(ctx, *vnic.PrivateIp, *vnic.SubnetId)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	glog.V(6).Infof("ipOcid: ", ipOcid)
+
+	for routeTableID := range routeTableIds {
+		routeTable, err := cp.client.Networking().GetRouteTable(ctx, routeTableID)
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		for _, oRoute := range routeTable.RouteRules {
+			if *oRoute.CidrBlock == route.DestinationCIDR {
+				if *oRoute.NetworkEntityId == ipOcid {
+					// skipping existing route table entry
+					glog.V(6).Info("Skipping existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					continue
+				} else {
+					oRoute.NetworkEntityId = &ipOcid
+					glog.V(6).Info("Update existing route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+					err := cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
+					if err != nil {
+						return errors.WithStack(err)
+					}
+					continue
+				}
+			}
+		}
+
+		// route rule doesn't exist in RouteTable
+		glog.V(6).Info("Add new route entry: ", route.DestinationCIDR, " -> ", ipOcid)
+		routeTable.RouteRules = append(routeTable.RouteRules, core.RouteRule{
+			CidrBlock:       &route.DestinationCIDR,
+			NetworkEntityId: &ipOcid,
+		})
+
+		err = cp.client.Networking().UpdateRouteTable(ctx, routeTableID, routeTable.RouteRules)
```

Comment:
You probably want to use ETAGs here otherwise you run the risk of concurrent updates causing conflicts and leaving things in an undesirable state.